astropy/cosmology/core.py

```diff
@@ -32,7 +35,140 @@ class CosmologyError(Exception):
     pass
 
 
-class Cosmology(metaclass=ABCMeta):
+class Parameter(property):
+    """Cosmological parameter.
+
+    Parameters
+    ----------
+    fget : callable or None, optional
+    fmt : str or None, optional
+        format specification.
+    doc : str or None, optional
+
+    Examples
+    --------
+    The most common use case of ``Parameter`` is to access the corresponding
+    private attribute.
+
+        >>> from astropy.cosmology import LambdaCDM
+        >>> from astropy.cosmology.core import Parameter
+        >>> class Example1(LambdaCDM):
+        ...     param = Parameter(doc="example parameter")
+        ...     def __init__(self, param=15):
+        ...         self._param = param
+        >>> Example1.param
+        <Parameter 'Example1.param' at ...
+        >>> Example1.param.default
+        15
+
+        >>> ex = Example1(param=12357)
+        >>> ex.param
+        12357
+
+    ``Parameter`` also supports custom ``getter`` methods.
+
+        >>> import astropy.units as u
+        >>> class Example2(LambdaCDM):
+        ...     param = Parameter(doc="example parameter")
+        ...     def __init__(self, param=15):
+        ...         self._param = param
+        ...     @param.getter
+        ...     def param(self):
+        ...         return self._param * u.m
+
+        >>> ex2 = Example2(param=12357)
+        >>> ex2.param
+        <Quantity 12357. m>
+
+    .. doctest::
+       :hide:
+
+       >>> from astropy.cosmology.core import _COSMOLOGY_CLASSES
+       >>> _ = _COSMOLOGY_CLASSES.pop(Example1.__qualname__)
+       >>> _ = _COSMOLOGY_CLASSES.pop(Example2.__qualname__)
+    """
+
+    def __init__(self, fget=None, fmt=".3g", doc=None):
+        super().__init__(fget=fget, doc=doc)
+        if doc is not None:
+            self.__doc__ = doc
+        self._fmt = str(fmt)
+
+    def __set_name__(self, objcls, name):
+        # attribute name
+        self._attr_name = name
+        self._attr_name_private = "_" + name
+
+        # default value (done here b/c 1st access to 'objcls')
+        nullparam = inspect.Parameter('None', 1, default=inspect._empty)
+        param = objcls._init_signature.parameters.get(name, nullparam)
+        self._default = param.default if param.default is not inspect._empty else None
+
+        # Register on the objcls, in order defined
+        if not hasattr(objcls, "__parameters__"):
+            objcls.__parameters__ = ()
+        objcls.__parameters__ += (name, )
```

Comment:
Thanks! I like the one-liner.
I'll look into using ``cls.__dict__`` again. In a previous implementation I was having difficulty with Parameter order, but numerous things have changed in the interim, so maybe ``cls.__dict__`` will work here.

Reply:
Followup: ``cls.__dict__`` has a few problems in ``__init_subclass__``. Most pertinently, inherited descriptors don't appear in the dict, only new descriptors.
The alternative, ``dir(cls)`` works, but does not maintain definition order.
Also, both `cls.__dict__`` and ``dir(cls)`` search through every attribute and method on the class. 
The solutions in ``data_info.py`` appear geared towards making instance descriptors... is there something specific?

I have figured out a solution to the ``__dict__`` issue:
```
parameters = [
            n for n in cls.__dict__.keys() | cls.__parameters__
            if (not n.startswith("_") and isinstance(getattr(cls, n, None), Parameter))]
```
but this likewise does not maintain definition order.

Reply:
Yes, I was just thinking that one could append new definitions to `cls.__parameters__`. In principle, those do not have to be checked to be of the right class.

Reply:
``FlatLambdaCDM`` removes the parameter ``Ode0``, so unfortunately I think all parameters need to be checked on every class definition.

Reply:
I'm confused: to remove a `Parameter` when one is a subclass is not really correct, since that means the subclass can no longer be used in place of the regular class (see https://en.wikipedia.org/wiki/Liskov_substitution_principle). Would it not be more reasonable to allow a parameter to be fixed in value?

Reply:
I get that, but I don't want to change long-standing API. `FlatLamdaCDM` does not accept `Ode0` as an input argument, while this is a mandatory argument for its parent class, `LamdaCDM`. For one it is an independent parameter, for the other a derived quantity.

Reply:
In `flatLamdaCDM` the `Ode0` Parameter is converted to a normal `property`. So instances are still interchangeable, even if the class instantiation process is not.

Reply:
OK, so that already does what I was suggesting - makes sense! So, I guess the issue is only about what and how to add to `__parameters__` - how about something like:
```
# Get parameters that are still defined as actual Parameters, either in this class or above.
parameters = [n for n in cls.__parameters__ if isinstance(getattr(cls, n), Parameter)]
# Add new parameter definitions
parameters += [n for n, v in cls.__dict__.items() if n not in parameters]
```

p.s. I had something similar in `data_info`, where status needed changing when things were overridden in the class definition. See
https://github.com/astropy/astropy/blob/0fd7ae818fed3abe4c468170a507d52ef91dc7e8/astropy/utils/data_info.py#L240-L258

Reply:
I use something similar to the above. Thanks!